File: src/services/learner-storage.ts
import { invoke } from "@tauri-apps/api/core";
import { isTauriContext } from "./tauri-bridge";
import type {
  LearnerProfile,
  CreateLearnerProfileData,
  LearnerMasteryData,
  ObjectiveMastery,
  QuickCheckResult,
} from "@/types";

// ============================================
// Profile Functions
// ============================================

/**
 * Get all learner profiles
 */
export async function getLearnerProfiles(): Promise<LearnerProfile[]> {
  if (!isTauriContext()) {
    // Browser fallback - use localStorage
    const stored = localStorage.getItem("learner-profiles");
    return stored ? JSON.parse(stored) : [];
  }

  const result = await invoke<string>("get_learner_profiles");
  return JSON.parse(result);
}

/**
 * Save a learner profile (create or update)
 */
export async function saveLearnerProfile(profile: LearnerProfile): Promise<void> {
  if (!isTauriContext()) {
    // Browser fallback - use localStorage
    const profiles = await getLearnerProfiles();
    const index = profiles.findIndex((p) => p.learnerId === profile.learnerId);
    if (index >= 0) {
      profiles[index] = profile;
    } else {
      profiles.push(profile);
    }
    localStorage.setItem("learner-profiles", JSON.stringify(profiles));
    return;
  }

  await invoke("save_learner_profile", {
    profile: JSON.stringify(profile),
  });
}

/**
 * Create a new learner profile
 */
export async function createLearnerProfile(
  data: CreateLearnerProfileData
): Promise<LearnerProfile> {
  const now = new Date().toISOString();
  const profile: LearnerProfile = {
    learnerId: crypto.randomUUID(),
    displayName: data.displayName,
    grade: data.grade,
    avatarEmoji: data.avatarEmoji || "🦁",
    preferences: {
      favoriteSubjects: data.preferences?.favoriteSubjects || [],
      sessionDuration: data.preferences?.sessionDuration || 30,
      visualLearner: data.preferences?.visualLearner ?? true,
    },
    adultConfidence: data.adultConfidence || "intermediate",
    createdAt: now,
    updatedAt: now,
  };

  await saveLearnerProfile(profile);
  return profile;
}

/**
 * Update an existing learner profile
 */
export async function updateLearnerProfile(
  learnerId: string,
  updates: Partial<Omit<LearnerProfile, "learnerId" | "createdAt">>
): Promise<LearnerProfile> {
  const profiles = await getLearnerProfiles();
  const profile = profiles.find((p) => p.learnerId === learnerId);

  if (!profile) {
    throw new Error(`Learner profile not found: ${learnerId}`);
  }

  const updated: LearnerProfile = {
    ...profile,
    ...updates,
    updatedAt: new Date().toISOString(),
  };

  await saveLearnerProfile(updated);
  return updated;
}

/**
 * Delete a learner profile and all associated data
 */
export async function deleteLearnerProfile(learnerId: string): Promise<void> {
  if (!isTauriContext()) {
    // Browser fallback
    const profiles = await getLearnerProfiles();
    const filtered = profiles.filter((p) => p.learnerId !== learnerId);
    localStorage.setItem("learner-profiles", JSON.stringify(filtered));
    localStorage.removeItem(`learner-mastery-${learnerId}`);
    localStorage.removeItem(`learner-quickchecks-${learnerId}`);
    return;
  }

  await invoke("delete_learner_profile", { learnerId });
}

// ============================================
// Mastery Functions
// ============================================

/**
 * Get mastery data for a learner
 */
export async function getLearnerMastery(learnerId: string): Promise<LearnerMasteryData> {
  if (!isTauriContext()) {
    // Browser fallback
    const stored = localStorage.getItem(`learner-mastery-${learnerId}`);
    if (stored) return JSON.parse(stored);
    return {
      learnerId,
      objectives: {},
      lastSessionDate: null,
    };
  }

  const result = await invoke<string>("get_learner_mastery", { learnerId });
  return JSON.parse(result);
}

/**
 * Save mastery for a single objective
 */
export async function saveObjectiveMastery(
  learnerId: string,
  objectiveMastery: ObjectiveMastery
): Promise<void> {
  if (!isTauriContext()) {
    // Browser fallback
    const mastery = await getLearnerMastery(learnerId);
    mastery.objectives[objectiveMastery.objectiveId] = objectiveMastery;
    mastery.lastSessionDate = new Date().toISOString();
    localStorage.setItem(`learner-mastery-${learnerId}`, JSON.stringify(mastery));
    return;
  }

  await invoke("save_objective_mastery", {
    learnerId,
    objectiveMastery: JSON.stringify(objectiveMastery),
  });
}

/**
 * Save complete mastery data (bulk update)
 */
export async function saveLearnerMastery(
  learnerId: string,
  masteryData: LearnerMasteryData
): Promise<void> {
  if (!isTauriContext()) {
    // Browser fallback
    localStorage.setItem(`learner-mastery-${learnerId}`, JSON.stringify(masteryData));
    return;
  }

  await invoke("save_learner_mastery", {
    learnerId,
    masteryData: JSON.stringify(masteryData),
  });
}

/**
 * Update mastery state for an objective (convenience function)
 */
export async function updateObjectiveMasteryState(
  learnerId: string,
  objectiveId: string,
  subject: string,
  state: ObjectiveMastery["state"],
  score?: number
): Promise<void> {
  const mastery = await getLearnerMastery(learnerId);
  const existing = mastery.objectives[objectiveId];

  const updated: ObjectiveMastery = {
    objectiveId,
    subject,
    state,
    lastScore: score ?? existing?.lastScore,
    attempts: (existing?.attempts || 0) + (score !== undefined ? 1 : 0),
    lastUpdated: new Date().toISOString(),
    notes: existing?.notes,
  };

  await saveObjectiveMastery(learnerId, updated);
}

// ============================================
// Quick Check Functions (Phase 2)
// ============================================

/**
 * Get quick check history for a learner
 */
export async function getQuickCheckHistory(
  learnerId: string,
  objectiveId?: string
): Promise<QuickCheckResult[]> {
  if (!isTauriContext()) {
    // Browser fallback
    const stored = localStorage.getItem(`learner-quickchecks-${learnerId}`);
    if (!stored) return [];
    const history: QuickCheckResult[] = JSON.parse(stored);
    if (objectiveId) {
      return history.filter((h) => h.objectiveId === objectiveId);
    }
    return history;
  }

  const result = await invoke<string>("get_quick_check_history", {
    learnerId,
    objectiveId: objectiveId || null,
  });
  return JSON.parse(result);
}

/**
 * Save a quick check result
 */
export async function saveQuickCheckResult(
  learnerId: string,
  result: Omit<QuickCheckResult, "resultId" | "createdAt">
): Promise<QuickCheckResult> {
  const fullResult: QuickCheckResult = {
    ...result,
    resultId: crypto.randomUUID(),
    createdAt: new Date().toISOString(),
  };

  if (!isTauriContext()) {
    // Browser fallback
    const history = await getQuickCheckHistory(learnerId);
    history.push(fullResult);
    localStorage.setItem(`learner-quickchecks-${learnerId}`, JSON.stringify(history));
    return fullResult;
  }

  await invoke("save_quick_check_result", {
    learnerId,
    result: JSON.stringify(fullResult),
  });

  return fullResult;
}

// ============================================
// Active Learner (localStorage only)
// ============================================

const ACTIVE_LEARNER_KEY = "active-learner-id";

/**
 * Get the active learner ID from localStorage
 */
export function getActiveLearnerIdFromStorage(): string | null {
  return localStorage.getItem(ACTIVE_LEARNER_KEY);
}

/**
 * Save the active learner ID to localStorage
 */
export function setActiveLearnerIdToStorage(learnerId: string | null): void {
  if (learnerId) {
    localStorage.setItem(ACTIVE_LEARNER_KEY, learnerId);
  } else {
    localStorage.removeItem(ACTIVE_LEARNER_KEY);
  }
}

File: src-tauri/src/commands/learner_storage.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::path::PathBuf;
use tokio::fs;

const LEARNERS_DIR: &str = "learners";
const PROFILES_FILE: &str = "profiles.json";
const MASTERY_FILE: &str = "mastery.json";

// Helper to get the learners directory
fn get_learners_dir(app_handle: &tauri::AppHandle) -> Result<PathBuf, String> {
    let app_data_dir = app_handle
        .path()
        .app_data_dir()
        .map_err(|e| format!("Failed to get app data directory: {}", e))?;
    Ok(app_data_dir.join(LEARNERS_DIR))
}

// Helper to get the profiles file path
fn get_profiles_path(app_handle: &tauri::AppHandle) -> Result<PathBuf, String> {
    Ok(get_learners_dir(app_handle)?.join(PROFILES_FILE))
}

// Helper to get a learner's data directory
fn get_learner_dir(app_handle: &tauri::AppHandle, learner_id: &str) -> Result<PathBuf, String> {
    Ok(get_learners_dir(app_handle)?.join(learner_id))
}

// ============================================
// Profile Commands
// ============================================

/// Get all learner profiles
#[tauri::command]
pub async fn get_learner_profiles(app_handle: tauri::AppHandle) -> Result<String, String> {
    let profiles_path = get_profiles_path(&app_handle)?;

    // If file doesn't exist, return empty array
    if !profiles_path.exists() {
        return Ok("[]".to_string());
    }

    fs::read_to_string(&profiles_path)
        .await
        .map_err(|e| format!("Failed to read profiles: {}", e))
}

/// Save a learner profile (upsert)
#[tauri::command]
pub async fn save_learner_profile(
    app_handle: tauri::AppHandle,
    profile: String,
) -> Result<(), String> {
    let learners_dir = get_learners_dir(&app_handle)?;
    let profiles_path = get_profiles_path(&app_handle)?;

    // Create directory if it doesn't exist
    fs::create_dir_all(&learners_dir)
        .await
        .map_err(|e| format!("Failed to create learners directory: {}", e))?;

    // Parse the incoming profile
    let new_profile: Value =
        serde_json::from_str(&profile).map_err(|e| format!("Invalid profile JSON: {}", e))?;

    let learner_id = new_profile
        .get("learnerId")
        .and_then(|v| v.as_str())
        .ok_or("Profile must have a learnerId")?;

    // Read existing profiles
    let mut profiles: Vec<Value> = if profiles_path.exists() {
        let content = fs::read_to_string(&profiles_path)
            .await
            .map_err(|e| format!("Failed to read profiles: {}", e))?;
        serde_json::from_str(&content).unwrap_or_else(|_| Vec::new())
    } else {
        Vec::new()
    };

    // Find and update existing profile, or add new one
    let mut found = false;
    for profile in profiles.iter_mut() {
        if profile.get("learnerId").and_then(|v| v.as_str()) == Some(learner_id) {
            *profile = new_profile.clone();
            found = true;
            break;
        }
    }
    if !found {
        profiles.push(new_profile.clone());
    }

    // Write profiles back
    let content = serde_json::to_string_pretty(&profiles)
        .map_err(|e| format!("Failed to serialize profiles: {}", e))?;
    fs::write(&profiles_path, content)
        .await
        .map_err(|e| format!("Failed to write profiles: {}", e))?;

    // Create learner directory
    let learner_dir = get_learner_dir(&app_handle, learner_id)?;
    fs::create_dir_all(&learner_dir)
        .await
        .map_err(|e| format!("Failed to create learner directory: {}", e))?;

    Ok(())
}

/// Delete a learner profile and all associated data
#[tauri::command]
pub async fn delete_learner_profile(
    app_handle: tauri::AppHandle,
    learner_id: String,
) -> Result<(), String> {
    let profiles_path = get_profiles_path(&app_handle)?;
    let learner_dir = get_learner_dir(&app_handle, &learner_id)?;

    // Remove from profiles list
    if profiles_path.exists() {
        let content = fs::read_to_string(&profiles_path)
            .await
            .map_err(|e| format!("Failed to read profiles: {}", e))?;
        let mut profiles: Vec<Value> =
            serde_json::from_str(&content).unwrap_or_else(|_| Vec::new());

        profiles.retain(|p| p.get("learnerId").and_then(|v| v.as_str()) != Some(&learner_id));

        let content = serde_json::to_string_pretty(&profiles)
            .map_err(|e| format!("Failed to serialize profiles: {}", e))?;
        fs::write(&profiles_path, content)
            .await
            .map_err(|e| format!("Failed to write profiles: {}", e))?;
    }

    // Delete learner directory and all contents
    if learner_dir.exists() {
        fs::remove_dir_all(&learner_dir)
            .await
            .map_err(|e| format!("Failed to delete learner data: {}", e))?;
    }

    Ok(())
}

// ============================================
// Mastery Commands
// ============================================

/// Get mastery data for a learner
#[tauri::command]
pub async fn get_learner_mastery(
    app_handle: tauri::AppHandle,
    learner_id: String,
) -> Result<String, String> {
    let learner_dir = get_learner_dir(&app_handle, &learner_id)?;
    let mastery_path = learner_dir.join(MASTERY_FILE);

    // If file doesn't exist, return default structure
    if !mastery_path.exists() {
        let default = serde_json::json!({
            "learnerId": learner_id,
            "objectives": {},
            "lastSessionDate": null
        });
        return Ok(default.to_string());
    }

    fs::read_to_string(&mastery_path)
        .await
        .map_err(|e| format!("Failed to read mastery data: {}", e))
}

/// Save mastery data for a specific objective
#[tauri::command]
pub async fn save_objective_mastery(
    app_handle: tauri::AppHandle,
    learner_id: String,
    objective_mastery: String,
) -> Result<(), String> {
    let learner_dir = get_learner_dir(&app_handle, &learner_id)?;
    let mastery_path = learner_dir.join(MASTERY_FILE);

    // Create directory if it doesn't exist
    fs::create_dir_all(&learner_dir)
        .await
        .map_err(|e| format!("Failed to create learner directory: {}", e))?;

    // Parse the incoming objective mastery
    let new_mastery: Value = serde_json::from_str(&objective_mastery)
        .map_err(|e| format!("Invalid mastery JSON: {}", e))?;

    let objective_id = new_mastery
        .get("objectiveId")
        .and_then(|v| v.as_str())
        .ok_or("Mastery must have an objectiveId")?;

    // Read existing mastery data or create default
    let mut mastery_data: Value = if mastery_path.exists() {
        let content = fs::read_to_string(&mastery_path)
            .await
            .map_err(|e| format!("Failed to read mastery data: {}", e))?;
        serde_json::from_str(&content).unwrap_or_else(|_| {
            serde_json::json!({
                "learnerId": learner_id,
                "objectives": {},
                "lastSessionDate": null
            })
        })
    } else {
        serde_json::json!({
            "learnerId": learner_id,
            "objectives": {},
            "lastSessionDate": null
        })
    };

    // Update the objectives map
    if let Some(objectives) = mastery_data.get_mut("objectives") {
        if let Some(obj_map) = objectives.as_object_mut() {
            obj_map.insert(objective_id.to_string(), new_mastery);
        }
    }

    // Update last session date
    let now = chrono::Utc::now().to_rfc3339();
    if let Some(obj) = mastery_data.as_object_mut() {
        obj.insert("lastSessionDate".to_string(), Value::String(now));
    }

    // Write mastery data back
    let content = serde_json::to_string_pretty(&mastery_data)
        .map_err(|e| format!("Failed to serialize mastery data: {}", e))?;
    fs::write(&mastery_path, content)
        .await
        .map_err(|e| format!("Failed to write mastery data: {}", e))?;

    Ok(())
}

/// Save complete mastery data for a learner (bulk update)
#[tauri::command]
pub async fn save_learner_mastery(
    app_handle: tauri::AppHandle,
    learner_id: String,
    mastery_data: String,
) -> Result<(), String> {
    let learner_dir = get_learner_dir(&app_handle, &learner_id)?;
    let mastery_path = learner_dir.join(MASTERY_FILE);

    // Create directory if it doesn't exist
    fs::create_dir_all(&learner_dir)
        .await
        .map_err(|e| format!("Failed to create learner directory: {}", e))?;

    // Validate JSON
    let _: Value =
        serde_json::from_str(&mastery_data).map_err(|e| format!("Invalid mastery JSON: {}", e))?;

    // Write mastery data
    fs::write(&mastery_path, &mastery_data)
        .await
        .map_err(|e| format!("Failed to write mastery data: {}", e))?;

    Ok(())
}

// ============================================
// Quick Check Commands (Phase 2)
// ============================================

const QUICK_CHECKS_FILE: &str = "quick-checks.json";

/// Get quick check history for a learner
#[tauri::command]
pub async fn get_quick_check_history(
    app_handle: tauri::AppHandle,
    learner_id: String,
    objective_id: Option<String>,
) -> Result<String, String> {
    let learner_dir = get_learner_dir(&app_handle, &learner_id)?;
    let checks_path = learner_dir.join(QUICK_CHECKS_FILE);

    // If file doesn't exist, return empty array
    if !checks_path.exists() {
        return Ok("[]".to_string());
    }

    let content = fs::read_to_string(&checks_path)
        .await
        .map_err(|e| format!("Failed to read quick check history: {}", e))?;

    // Filter by objective_id if provided
    if let Some(obj_id) = objective_id {
        let checks: Vec<Value> = serde_json::from_str(&content).unwrap_or_else(|_| Vec::new());
        let filtered: Vec<&Value> = checks
            .iter()
            .filter(|c| c.get("objectiveId").and_then(|v| v.as_str()) == Some(&obj_id))
            .collect();
        return serde_json::to_string(&filtered)
            .map_err(|e| format!("Failed to serialize filtered history: {}", e));
    }

    Ok(content)
}

/// Save a quick check result
#[tauri::command]
pub async fn save_quick_check_result(
    app_handle: tauri::AppHandle,
    learner_id: String,
    result: String,
) -> Result<(), String> {
    let learner_dir = get_learner_dir(&app_handle, &learner_id)?;
    let checks_path = learner_dir.join(QUICK_CHECKS_FILE);

    // Create directory if it doesn't exist
    fs::create_dir_all(&learner_dir)
        .await
        .map_err(|e| format!("Failed to create learner directory: {}", e))?;

    // Parse the incoming result
    let new_result: Value =
        serde_json::from_str(&result).map_err(|e| format!("Invalid result JSON: {}", e))?;

    // Read existing history
    let mut history: Vec<Value> = if checks_path.exists() {
        let content = fs::read_to_string(&checks_path)
            .await
            .map_err(|e| format!("Failed to read quick check history: {}", e))?;
        serde_json::from_str(&content).unwrap_or_else(|_| Vec::new())
    } else {
        Vec::new()
    };

    // Add new result
    history.push(new_result);

    // Write history back
    let content = serde_json::to_string_pretty(&history)
        .map_err(|e| format!("Failed to serialize history: {}", e))?;
    fs::write(&checks_path, content)
        .await
        .map_err(|e| format!("Failed to write quick check history: {}", e))?;

    Ok(())
}
